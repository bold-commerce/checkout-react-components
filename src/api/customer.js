import fetchApi from '../utils/fetchApi';

export const validateCustomer = async (csrf, apiPath, customer) => {
  const email = encodeURIComponent(customer.email_address);

  const response = await fetchApi(`${apiPath}/validate_email_address?email_address=${email}`, {
    headers: {
      'Content-Type': 'application/json',
      'X-CSRF-TOKEN': csrf,
    },
  });

  return response;
};

export const updateCustomer = async (csrf, apiPath, customer, method) => {
  const response = await fetchApi(`${apiPath}/customer/guest`, {
    method,
    headers: {
      'Content-Type': 'application/json',
      'X-CSRF-TOKEN': csrf,
    },
    body: JSON.stringify(customer),
  });

  return response;
};
