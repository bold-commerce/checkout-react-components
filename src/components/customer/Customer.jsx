/* eslint-disable react/forbid-prop-types */
import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { InputField } from '@boldcommerce/stacks-ui';
import useCustomer from '../../hooks/useCustomer';
import './Customer.css';

export const Customer = ({
  customer, customerErrors, isAuthenticated, submitCustomer, onChange, onLogin,
}) => {
  const [email, setEmail] = useState(customer?.email_address);

  useEffect(() => {
    if (onChange) {
      onChange({
        email_address: email,
      });
    }
  }, [email]);

  return (
    <section className="FieldSet FieldSet--CustomerInformation">
      <div className="FieldSet__Header">
        <div className="FieldSet__Heading">Customer information</div>
        {
          onLogin && (
            <p className="FieldSet__SubHeading">
              Already have an account with us?
              {' '}
              <button type="button" onClick={onLogin}>Log in</button>
            </p>
          )
        }
      </div>
      <div className="FieldSet__Content">
        <InputField
          className="InputField Field--Email"
          placeholder="Email"
          type="email"
          name="email"
          messageType={customerErrors && 'alert'}
          messageText={customerErrors && customerErrors.email}
          value={email}
          onChange={(e) => setEmail(e.target.value)}
          readOnly={isAuthenticated}
          onBlur={() => submitCustomer({
            email_address: email,
          })}
        />
      </div>
    </section>
  );
};

Customer.propTypes = {
  customer: PropTypes.object,
  customerErrors: PropTypes.object,
  isAuthenticated: PropTypes.bool,
  submitCustomer: PropTypes.func,
  onChange: PropTypes.func,
  onLogin: PropTypes.func,
};

const MemoizedCustomer = React.memo(Customer);

/**
 * If onChange is provided, stop automatic submitting of customer to api and instead pass customer object up to parent through onChange callback
 */
const CustomerContainer = ({ onChange, onLogin }) => {
  const {
    customer, customerErrors, isAuthenticated, submitCustomer,
  } = useCustomer();

  return (
    <MemoizedCustomer
      customer={customer}
      customerErrors={customerErrors}
      isAuthenticated={isAuthenticated}
      onChange={onChange}
      submitCustomer={onChange || submitCustomer}
      onLogin={onLogin}
    />
  );
};

CustomerContainer.propTypes = {
  onChange: PropTypes.func,
  onLogin: PropTypes.func,
};

export default CustomerContainer;
