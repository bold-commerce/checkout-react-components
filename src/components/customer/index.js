// eslint-disable-next-line import/prefer-default-export
export { default as Customer } from './Customer';
