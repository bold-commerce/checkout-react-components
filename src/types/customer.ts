import { Address } from "./Address";

export interface Customer {
  platform_id: string,
  public_id: string,
  first_name: string,
  last_name: string,
  email_address: string,
  accepts_marketing: boolean,
  saved_addresses: Address[],
} 