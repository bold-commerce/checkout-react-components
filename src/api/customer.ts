import { Customer } from '../types/Customer';
import fetchApi from '../utils/fetchApi';

export const validateCustomer = async (token: string, apiPath: string, customer: Customer) => {
  const email = encodeURIComponent(customer.email_address);

  const response = await fetchApi(`${apiPath}/validate_email_address?email_address=${email}`, {
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${token}`,
    },
  });

  return response;
};

export const updateCustomer = async (token: string, apiPath: string, customer: Customer, method: "PUT" | "POST") => {
  const response = await fetchApi(`${apiPath}/customer/guest`, {
    method,
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${token}`,
    },
    body: JSON.stringify(customer),
  });

  return response;
};
