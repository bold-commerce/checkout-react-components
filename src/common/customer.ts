import { IApiReturnObject, updateCustomer } from '@bold-commerce/checkout-frontend-library';
import React from 'react';
import { DispatchActions, IState } from 'src/store';
import { IUpdateCustomerMethod } from 'src/types';

export const updateCustomerMethod = (dispatch: React.Dispatch<DispatchActions>, getState: () => IState) => async (data: IUpdateCustomerMethod): Promise<IApiReturnObject> => {
    const {
        email_address: email,
        first_name: firstName,
        last_name: lastName,
        accepts_marketing: acceptsMarketing,
    } = data;

    // eslint-disable-next-line no-console
    console.log(getState());
    
    // TODO: Change between add guest customer and update customer
    const response = await updateCustomer(firstName || '', lastName || '', email || '', acceptsMarketing || false);

    return response;
};
