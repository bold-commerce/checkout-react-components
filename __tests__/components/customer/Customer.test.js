import React from 'react';
import { render } from '@testing-library/react';
import { Customer } from '../../../src/components/customer/Customer';

describe('Customer', () => {
  test('renders Customer component', () => {
    const { asFragment } = render(<Customer customer={null} customerErrors={null} isAuthenticated={false} submitCustomer={null} onChange={null} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders Customer component with errors', () => {
    const errors = {
      email: 'Email format is invalid.',
    };

    const { asFragment } = render(<Customer customer={null} customerErrors={errors} isAuthenticated={false} submitCustomer={null} onChange={null} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders authenticated Customer with pre-filled data', () => {
    const customer = {
      email_address: 'john.doe@email.com',
    };

    const { asFragment } = render(<Customer customer={customer} customerErrors={null} isAuthenticated submitCustomer={null} onChange={null} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
