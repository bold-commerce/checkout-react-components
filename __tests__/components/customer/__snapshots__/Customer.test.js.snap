// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Customer renders Customer component 1`] = `
<DocumentFragment>
  <section
    class="FieldSet FieldSet--CustomerInformation"
  >
    <div
      class="FieldSet__Header"
    >
      <div
        class="FieldSet__Heading"
      >
        Customer information
      </div>
    </div>
    <div
      class="FieldSet__Content"
    >
      <div
        class="stx-field stx-field--with-input InputField Field--Email"
      >
        <div
          class="stx-input"
        >
          <input
            aria-readonly="false"
            class="stx-input__input-element"
            id="stxField1"
            name="email"
            placeholder="Email"
            type="email"
            value=""
          />
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`Customer renders Customer component with errors 1`] = `
<DocumentFragment>
  <section
    class="FieldSet FieldSet--CustomerInformation"
  >
    <div
      class="FieldSet__Header"
    >
      <div
        class="FieldSet__Heading"
      >
        Customer information
      </div>
    </div>
    <div
      class="FieldSet__Content"
    >
      <div
        class="stx-field stx-field--alert stx-field--with-input InputField Field--Email"
      >
        <div
          class="stx-input stx-input--alert"
        >
          <input
            aria-readonly="false"
            class="stx-input__input-element"
            id="stxField2"
            name="email"
            placeholder="Email"
            type="email"
            value=""
          />
        </div>
        <div
          class="stx-field__message"
        >
          Email format is invalid.
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`Customer renders authenticated Customer with pre-filled data 1`] = `
<DocumentFragment>
  <section
    class="FieldSet FieldSet--CustomerInformation"
  >
    <div
      class="FieldSet__Header"
    >
      <div
        class="FieldSet__Heading"
      >
        Customer information
      </div>
    </div>
    <div
      class="FieldSet__Content"
    >
      <div
        class="stx-field stx-field--with-input InputField Field--Email"
        readonly=""
      >
        <div
          class="stx-input stx-input--is-read-only"
        >
          <input
            aria-readonly="true"
            class="stx-input__input-element"
            id="stxField3"
            name="email"
            placeholder="Email"
            readonly=""
            type="email"
            value="john.doe@email.com"
          />
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
